### ISSUE TITLE
[question] search for packages based on build info

### ISSUE THREAD

**Issue by tbsuht** [conan-io/conan#15411]
### What is your question?

Hi,

this is a follow-up question of https://github.com/conan-io/conan/issues/14464 where a solution has been merged.

I'm not sure how to use the already merged stuff, so here is what I want to do:

I'm iterating the build output and for some dependencies in the info I want to search for a `build_type=Debug` package in the local cache with the same settings and options used during the build.

In the past I did something like this:
```
packages = conan_api.list.filter_packages_configurations(conan_api.list.packages_configurations(RecipeReference.loads(ref)),
                                                                                                                create_debug_filter(dependency))
```
where `create_debug_filter` was just creating a query based on the settings and options in the build info and overwrote the `build_type` setting.

Now with the new API I guess I can use
```
conan_api.list.explain_missing_binaries(ref, conaninfo, remotes)
```

but I'm not sure how to fill the conaninfo in my case? How to get a conaninfo based on a package-ref?

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @tbsuht 

Thanks for your question.

> I'm iterating the build output and for some dependencies in the info I want to search for a build_type=Debug package in the local cache with the same settings and options used during the build.

What is the problem that you are trying to solve? are you sure that you don't want to define a ``compatibility.py`` to automatically fallback and use Debug packages if available (or the opposite)?


> conan_api.list.explain_missing_binaries(ref, conaninfo, remotes)

the new API doesn't seem intended for your use case, but for a user perspective to understand what could be the differences generating a "missing binary" message. Maybe you just want to keep using the other API?

**Comment by tbsuht** [comment]
Hi @memsharded 

> Maybe you just want to keep using the other API?

The other API is not working as intended, this is due to what is described in the first post in https://github.com/conan-io/conan/issues/14464. The query must omit parameters which are not part of the package id and currently I can only achieve this by defining those parameters manually.

> What is the problem that you are trying to solve? are you sure that you don't want to define a compatibility.py to automatically fallback and use Debug packages if available (or the opposite)?

I have created some scripts which help to remotely debug code on our embedded devices. 
Especially uploading the required code: the developer can select in a json file which dependencies should be uploaded in debug format.

In order to do that, I need to check before starting a debugging session, whether there is already a debug package available which was build locally.

**Comment by memsharded** [comment]
Why not iterating directly the packages returned from ``packages = conan_api.list.packages_configurations(RecipeReference.loads(ref))``? You might iterate them ``for pref, conf in packages.items()``, and for each one explore its ``conf``. That would be a relatively simple dictionary containing the different binary info (``conaninfo.txt``) sections like ``requires``, ``settings``, ``options``. Have you considered this approach? I think it might be able to implement the custom logic you are looking for?

**Comment by tbsuht** [comment]
No, I didn't consider it yet. This seems possible and probably a better solution than creating a query for `filter_packages_configurations`.

But: this is too complicated :) 
Isn't it possible to provide a profile to `filter_packages_configurations` or `packages_configurations`?

**Comment by tbsuht** [comment]
> Why not iterating directly the packages returned from `packages = conan_api.list.packages_configurations(RecipeReference.loads(ref))`? You might iterate them `for pref, conf in packages.items()`, and for each one explore its `conf`. That would be a relatively simple dictionary containing the different binary info (`conaninfo.txt`) sections like `requires`, `settings`, `options`. Have you considered this approach? I think it might be able to implement the custom logic you are looking for?

Ok, I tried this and this has nearly the same problem:

`conan_api.list.packages_configurations` will return confs *without* options that are not part of the package id.
Example:
```
{'build_gmock': 'True', 'disable_pthreads': 'False', 'hide_symbols': 'False', 'shared': 'True'}
```

But if I read the confs from the build info, those options are present...
```
{'build_gmock': 'True', 'disable_pthreads': 'False', 'hide_symbols': 'False', 'no_main': 'False', 'shared': 'True'}
```

The only way I could think of is ignoring options which are part of the build info but not returned by `conan_api.list.packages_configurations`. Is this safe to do?

**Comment by memsharded** [comment]
> conan_api.list.packages_configurations will return confs without options that are not part of the package id.

Yes, this is by design, the package configuraitons are the ``package-ids`` that are computed from ``conaninfo.txt`` that only contains the information relevant to the ``package_id``

> But if I read the confs from the build info,

Sorry, I am a bit confused here, where are your reading the confs from? what is that ``build info``?

> In order to do that, I need to check before starting a debugging session, whether there is already a debug package available which was build locally.

Maybe the issue is that I don't fully understand the flow. If the developers want to debug some dependencies, and they select it somehow in a json, why a ``conan install`` of those dependencies with the ``-s <mydep>/*:build_type=Debug`` (iterating for the developer defined dependencies) doesn't work? That will automatically check for the ``Debug`` packages for those dependencies. Why trying to find them via ``conan list`` when ``conan install`` or ``conan graph info`` will compute things "exactly" and will directly report what is missing and what not?

**Comment by tbsuht** [comment]
> Sorry, I am a bit confused here, where are your reading the confs from? what is that build info?

What I mean is the output of "conan install" for the project I want to build. Maybe `install info` is a better wording, idk.

> Maybe the issue is that I don't fully understand the flow

Ok, so this is our current flow: we build our projects and dependencies within CI with `build_type=Release` and also `build_type=Debug`. The packages are stored in artifactory, so that developers can build projects very fast locally if they want to change something.

For debugging into code we stick to this guide: https://docs.conan.io/1/using_packages/debugging.html
So if a developer wants to debug into a specific package and step through the code, he will need to rebuild the package locally.

So locally inside a project there is a small json file which is automatically generated by our debug scripts, listing the dependencies, which looks like this:
```
{
    "libbsd": "y",
    "gtest": "y",
     ...
}
```

"y" means: I want to upload this library to the remote target and possibly step through the code.
"n" means: I don't want to upload this library, it is already part of the remote device.

So if the setting is "y", we need a package with `build_type=Debug` which was locally build. The Debug package from the CI can't be used in this case, because the debugger will not be able to locate the sources.

> If the developers want to debug some dependencies, and they select it somehow in a json, why a conan install of those dependencies with the -s <mydep>/*:build_type=Debug (iterating for the developer defined dependencies) doesn't work?

How to make sure that those dependencies are only build once locally and not pulled from artifactory (as there are Debug packages stored there)? Currently I'm using `conan_api.cache.build_path` to check whether the package was indeed build locally.

**Comment by memsharded** [comment]
Ok, that is helpful information.
I am trying to think a bit outside the box, and trying to use some more built-in mechanism. Consider this test:
```python
def test_build_local_debug():
    c = TestClient()
    settings_user = textwrap.dedent("""\
        build_type:
            Release:
            Debug:
                local: [null, True]
            """)
    save(c.cache.settings_path_user, settings_user)
    c.save({"conanfile.py": GenConanfile("pkg", "0.1").with_settings("os", "build_type")})
    c.run("create . -s os=Windows -s build_type=Debug")

    c.run("install --requires=pkg/0.1 "
          "-s os=Windows -s build_type=Debug -s build_type.local=True", assert_error=True)
    # It fails, it doesn't find the binary because it was not built locally
    assert "ERROR: Missing prebuilt package for 'pkg/0.1'" in c.out
    # this will build the local Debug binary for ``pkg`` only if it was not built before
    c.run("install --requires=pkg/0.1 --build=missing "
          "-s os=Windows -s build_type=Debug -s pkg/*:build_type.local=True")
    print(c.out)
    c.run("list *:*")  # Check that we have 2 binaries, teh "server" one and the one built locally
    print(c.out)
    c.run("remove *:* -p build_type.local=True -c")
    print(c.out)
    c.run("list *:*")  # Make sure we removed the build_type.local=True
    print(c.out)
```
- I am extending the settings via a simple ``settings_user.yml`` that extends the ``Debug`` build-type, to add it a "local" specifier
- I can control with the ``-s pkgname/*:build_type.local=True`` which packages I want to force to be built locally
- I can easily remove the ``build_type.local=True`` packages from local with a single command (in general it might not be necessary, as CI will not use this, only developers)

With this approach you would only have to care to translate your json inputs to ``-s pkg1/*:build_type.local=True -s pkg2/*:build_type.local=True`` arguments. No need for any custom logic in recipes or anywhere else.

**Comment by tbsuht** [comment]
That's actually a nice solution, thanks for showing this!

The problem for me is: non of the developers in my division wants to deal with working to much with a terminal or with conan itself. Currently building and debugging is a one-click task, hiding everything from the developer via cmake_conan.

While probably being the cleaner solution, I think this requires quite a few adjustments (cmake_conan and settings_user at least). I'll keep it in the back of my head, thanks!
